File: 658.py
class Solution:
    def findClosestElements(self, arr: list[int], k: int, x: int) -> list[int]:
        if k == len(arr):
            return arr
        ans = []
        # step 1: find the "closest" elt
        idxOfClosest = -1
        valOfClosest = 10**5
        for i in range(len(arr)):
            if valOfClosest > abs(arr[i] - x):
                valOfClosest = abs(arr[i] - x)
                idxOfClosest = i
        l, r = idxOfClosest - 1, idxOfClosest + 1
        ans.append(arr[idxOfClosest])
        k -= 1
        while (l > -1 or r < len(arr)) and k != 0:
            lVal = False
            rVal = False
            if l > -1:
                lVal = abs(arr[l] - x)
            if r < len(arr):
                rVal = abs(arr[r] - x)
            # find which one of l,r is "closer" and add that one to ans
            if lVal is False:
                ans.append(arr[r])
                r += 1
            elif rVal is False:
                ans.append(arr[l])
                l -= 1
            elif rVal < lVal:
                ans.append(arr[r])
                r += 1
            else:
                ans.append(arr[l])
                l -= 1
            k -= 1
        ans.sort()
        print(ans)
        return ans


# Solution().findClosestElements([1, 2, 3, 4, 5], 4, 3)
Solution().findClosestElements([1, 2, 3, 4, 5], 4, -1)
# Solution().findClosestElements([1, 1, 1, 1, 2], 4, 3)
# Solution().findClosestElements([0, 0, 0, 1, 3, 5, 6, 7, 8, 8], 2, 2)

File: 658.cpp
#include <algorithm>
#include <iostream>
#include <math.h>
#include <vector>
using namespace std;

class Solution {
private:
  struct p {
    int val;
    bool valid;
  };

public:
  vector<int> findClosestElements(vector<int> &arr, int k, int x) {
    if (k == arr.size()) {
      return arr;
    }
    vector<int> ans;
    int idxOfMidpoint = -1;
    int bestValSoFar = 10e5 + 1;
    for (int i = arr.size() - 1; i > -1; --i) {
      if (abs(arr[i] - x) <= bestValSoFar) {
        bestValSoFar = abs(arr[i] - x);
        idxOfMidpoint = i;
      }
    }
    --k;
    ans.push_back(arr[idxOfMidpoint]);
    int l = idxOfMidpoint - 1, r = idxOfMidpoint + 1;
    while ((l > -1 || r < arr.size()) && k != 0) {
      p lp = {-1, false};
      p rp = {-1, false};
      if (l > -1) {
        lp.valid = true;
        lp.val = abs(arr[l] - x);
      }
      if (r < arr.size()) {
        rp.valid = true;
        rp.val = abs(arr[r] - x);
      }
      if (!lp.valid) {
        ans.push_back(arr[r]);
        ++r;
      } else if (!rp.valid) {
        ans.push_back(arr[l]);
        --l;
      } else if (rp.val < lp.val) {
        ans.push_back(arr[r]);
        ++r;
      } else {
        ans.push_back(arr[l]);
        --l;
      }
      --k;
    }
    sort(ans.begin(), ans.end());
    // for (auto i : ans) {
    //   cout << i << " ";
    // }
    // cout << "\n";
    return ans;
  }
};

int main() {
  Solution s;
  vector<int> t1_in = {1, 3};
  s.findClosestElements(t1_in, 1, 2);
}